session.go

```diff
@@ -302,9 +305,18 @@ runLoop:
 		if !s.receivedTooManyUndecrytablePacketsTime.IsZero() && s.receivedTooManyUndecrytablePacketsTime.Add(protocol.PublicResetTimeout).Before(now) && len(s.undecryptablePackets) != 0 {
 			s.closeLocal(qerr.Error(qerr.DecryptionFailure, "too many undecryptable packets received"))
 		}
+
 		if now.Sub(s.lastNetworkActivityTime) >= s.idleTimeout() {
 			s.closeLocal(qerr.Error(qerr.NetworkIdleTimeout, "No recent network activity."))
+		} else if s.config.Keepalive {
+			timeElapsed := now.Sub(s.lastNetworkActivityTime)
+			if timeElapsed >= s.idleTimeout()/2 {
+				// send the PING frame since there is no activity in the session
+				s.packer.QueueControlFrameForNextPacket(&frames.PingFrame{})
+				s.keepalivePingSent = true
```

Comment:
I just realized this code block is below where we try to send, so we have to call `s.scheduleSending()` here so that the PING frame will be sent asap.